Websocket timeout

I've seen this exception more than once:

```java
SaltyRTC.RSigna: A WebSocket connect error occured: Failed to connect to '<domain>:443': failed to connect to <domain>/<ipv6> (port 443) after 2000ms                 
                 com.neovisionaries.ws.client.WebSocketException: Failed to connect to '<domain>:443': failed to connect to <domain>/<ipv6> (port 443) after 2000ms
                     at com.neovisionaries.ws.client.SocketConnector.doConnect(SocketConnector.java:119)
                     at com.neovisionaries.ws.client.SocketConnector.connect(SocketConnector.java:81)
                     at com.neovisionaries.ws.client.WebSocket.connect(WebSocket.java:2031)
                     at com.neovisionaries.ws.client.ConnectThread.run(ConnectThread.java:37)
                  Caused by: java.net.SocketTimeoutException: failed to connect to <domain>/<ipv6> (port 443) after 2000ms
                     at libcore.io.IoBridge.connectErrno(IoBridge.java:169)
                     at libcore.io.IoBridge.connect(IoBridge.java:122)
                     at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:183)
                     at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:456)
                     at java.net.Socket.connect(Socket.java:882)
                     at com.neovisionaries.ws.client.SocketConnector.doConnect(SocketConnector.java:110)
                     at com.neovisionaries.ws.client.SocketConnector.connect(SocketConnector.java:81) 
                     at com.neovisionaries.ws.client.WebSocket.connect(WebSocket.java:2031) 
                     at com.neovisionaries.ws.client.ConnectThread.run(ConnectThread.java:37) 
SaltyRTC.RSigna: Could not connect to websocket: Failed to connect to '<domain>:443': failed to connect to <domain>/<ipv6> (port 443) after 2000ms
```

- Timeout should be longer than 2000ms by default
- Timeout should be configurable by the user application
- Maybe add a retry, or at least pass the failed state up to the user application

--- a/src/main/java/org/saltyrtc/client/SaltyRTC.java
+++ b/src/main/java/org/saltyrtc/client/SaltyRTC.java
@@ -131,7 +131,7 @@ public class SaltyRTC {
      * Connect asynchronously to the SaltyRTC server.
      *
      * To get notified when the connection is up and running,
-     * subscribe to the `ConnectedEvent`.
+     * subscribe to the `SignalingStateChangedEvent`.
      *
      * @throws ConnectionException if setting up the WebSocket connection fails.
      */

--- a/src/main/java/org/saltyrtc/client/signaling/Signaling.java
+++ b/src/main/java/org/saltyrtc/client/signaling/Signaling.java
@@ -94,6 +94,7 @@ public abstract class Signaling implements SignalingInterface {
     private WebSocket ws;
     private int pingInterval;
     private int wsConnectTimeout;
+    private int wsConnectAttempt = 0;
 
     // Connection state
     private SignalingState state = SignalingState.NEW;
@@ -210,7 +211,9 @@ public abstract class Signaling implements SignalingInterface {
     /**
      * Connect asynchronously to the SaltyRTC server.
      *
-     * To get notified when the connection is up and running, subscribe to the `ConnectedEvent`.
+     * To get notified when the connection is up and running, subscribe to the `SignalingStateChangedEvent`.
+     *
+     * @throws ConnectionException if setting up the WebSocket connection fails.
      */
     public void connect() throws ConnectionException {
         this.getLogger().info("Connecting to SaltyRTC server at "
@@ -219,7 +222,7 @@ public abstract class Signaling implements SignalingInterface {
         try {
             this.initWebsocket();
         } catch (IOException e) {
-            throw new ConnectionException("Connecting to WebSocket failed.", e);
+            throw new ConnectionException("Setting up WebSocket failed.", e);
         }
         this.connectWebsocket();
     }
@@ -311,8 +314,16 @@ public abstract class Signaling implements SignalingInterface {
             @Override
             @SuppressWarnings("UnqualifiedMethodAccess")
             public void onConnectError(WebSocket websocket, WebSocketException ex) throws Exception {
-                getLogger().error("Could not connect to websocket: " + ex.getMessage());
-                setState(SignalingState.ERROR);
+                getLogger().error("Could not connect to websocket (" + ex.getError().toString() + "): " + ex.getMessage());
+                if (Signaling.this.wsConnectAttempt == 1) {
+                    getLogger().info("Retrying to reconnect once...");
+                    Signaling.this.wsConnectAttempt += 1;
+                    Signaling.this.setState(SignalingState.WS_CONNECTING);
+                    Signaling.this.ws.recreate(Signaling.this.wsConnectTimeout).connectAsynchronously();
+                } else {
+                    getLogger().info("Giving up.");
+                    setState(SignalingState.ERROR);
+                }
             }
 
             @Override
@@ -456,11 +467,16 @@ public abstract class Signaling implements SignalingInterface {
                 }
             }
 
+            /**
+             * WebSocketListener has some onXxxError() methods such as onFrameError() and onSendError(). Among such
+             * methods, onError() is a special one. It is always called before any other onXxxError() is called.
+             * For example, in the implementation of run() method of ReadingThread, Throwable is caught and onError()
+             * and onUnexpectedError() are called in this order. The following is the implementation.
+             */
             @Override
             @SuppressWarnings("UnqualifiedMethodAccess")
             public void onError(WebSocket websocket, WebSocketException cause) throws Exception {
-                getLogger().error("A WebSocket connect error occured: " + cause.getMessage(), cause);
-                // TODO: Do we need to handle these?
+                getLogger().warn("A WebSocket error occured: " + cause.getMessage(), cause);
             }
 
             @Override
@@ -480,7 +496,6 @@ public abstract class Signaling implements SignalingInterface {
                 .setPingInterval(SALTYRTC_WS_PING_INTERVAL)
                 .addProtocol(SALTYRTC_SUBPROTOCOL)
                 .addListener(listener);
-
     }
 
     /**
@@ -488,6 +503,7 @@ public abstract class Signaling implements SignalingInterface {
      */
     private void connectWebsocket() {
         this.setState(SignalingState.WS_CONNECTING);
+        this.wsConnectAttempt = 1;
         this.ws.connectAsynchronously();
     }
 

